server/src/main/scala/org/http4s/server/middleware/CORS.scala

```diff
@@ -54,19 +55,28 @@ object CORS {
         else
           config.exposedHeaders.map(headerFromStrings("Access-Control-Expose-Headers", _))
 
+      def checkVaryHeader(response: Response[G]): Option[Header] = {
```

Comment:
I think this function can be `Response[G] => Response[G]`, saving you a fold down the line

If you look down below:

```scala
checkVaryHeader(withMethodBasedHeader)
          .fold(withMethodBasedHeader)(h => withMethodBasedHeader.putHeaders(h))
```

you could simplify this into:

```scala
def varyHeader(response: Response[G]): Response[G] =  {
  // Match avoids a fold higher order call, but you can just do this with fold if you don't want
  response.headers.get(CaseInsensitiveString("Vary")) match {
    case None => 
      response.putHeaders(Header("Vary", "Origin,Access-Control-Request-Method"))
     case _ => 
       response
  }
}
```

and then, down the line:

```scala
varyHeader(withMethodBasedHeader)
  .putHeaders(..
```